# DSE 310 Assignment 2
## Rita Abani
## Roll no: 19244
### Department of Electrical Engineering and Computer Science, IISER Bhopal

Throughout this code file, we are going to use the mysql.connector, hence we first install the mysql package in python.
NO MYSQL commands have been used, only Python commands have been used in this notebook after establishing database connectivity with MySQL from within this python notebook itself.

In [2]:
pip install mysql

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'C:\Users\DRA\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip' command.


In [1]:
import mysql.connector

 ### Part (a) Do data definition operations (Create, Drop and Modify schemas/tables)

 The necessary terms have been commented. We start by establishing connection, creating a cursor object and then preparing a query to create a database ' DSE 310Rita' that we are going to be using throughout this assignment. As and when required, to complete the tasks, tables have been added and modified in the same database.

In [5]:
#establishing the connection
conn = mysql.connector.connect(user='root', password='ainucurie@22', host='127.0.0.1')

#Creating a cursor object using the cursor() method
cursor = conn.cursor()
#Doping database MYDATABASE if already exists.
cursor.execute("DROP database IF EXISTS DSE310Rita")

#Preparing query to create a database
sql = "CREATE database DSE310Rita";

#Creating a database
cursor.execute(sql)

#Retrieving the list of databases
print("List of databases: ")
cursor.execute("SHOW DATABASES")
print(cursor.fetchall())


#Executing an MYSQL function using the execute() method
cursor.execute("SELECT DATABASE()")

# Fetch a single row using fetchone() method.
data = cursor.fetchone()
print("Connection established to: ",data)

#Closing the connection
conn.close()

List of databases: 
[('dbms_coursework',), ('dse310rita',), ('information_schema',), ('mysql',), ('performance_schema',), ('sakila',), ('sys',), ('world',)]
Connection established to:  (None,)


##### Creating the table using Python

In [3]:


#establishing the connection
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)

#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Dropping EMPLOYEE table if already exists.
cursor.execute("DROP TABLE IF EXISTS EMPLOYEE")

#Creating table as per requirement
sql ='''CREATE TABLE EMPLOYEE(
   FIRST_NAME CHAR(20) NOT NULL,
   LAST_NAME CHAR(20),
   AGE INT,
   SEX CHAR(1),
   INCOME FLOAT
)'''
cursor.execute(sql)

#Closing the connection
#conn.close()

  ####  Task (b) Do data manipulation operations (Insert, Update, Select, )

##### We are now inserting sample data into the table

In [4]:
#establishing the connection
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

# Preparing SQL query to INSERT a record into the database.
sql = """INSERT INTO EMPLOYEE(
   FIRST_NAME, LAST_NAME, AGE, SEX, INCOME)
   VALUES ('Mac', 'Mohan', 20, 'M', 2000)"""

try:
   # Executing the SQL command
   cursor.execute(sql)

   # Commit your changes in the database
   conn.commit()

except:
   # Rolling back in case of error
   conn.rollback()

# Closing the connection
conn.close()

##### Selecting information from the employee table in Python

In [9]:

#restablishing the connection
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Retrieving single row
sql = '''SELECT * from EMPLOYEE'''

#Executing the query
cursor.execute(sql)

#Fetching 1st row from the table
result = cursor.fetchone();
print(result)

#Fetching 1st row from the table
result = cursor.fetchall();
print(result)

#Closing the connection
conn.close()

('Mac', 'Mohan', 20, 'M', 2000.0)
[]


##### Using the Update statement in Python

In [2]:
#establishing the connection
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Dropping EMPLOYEE table if already exists.
cursor.execute("DROP TABLE IF EXISTS EMPLOYEE")
sql = '''CREATE TABLE EMPLOYEE(
   FIRST_NAME CHAR(20) NOT NULL,
   LAST_NAME CHAR(20),
   AGE INT,
   SEX CHAR(1),
   INCOME FLOAT
)'''
cursor.execute(sql)

#Inserting a few more records to the table
insert_stmt = "INSERT INTO EMPLOYEE (FIRST_NAME, LAST_NAME, AGE, SEX, INCOME) VALUES (%s, %s, %s, %s, %s)"

data = [('Krishna', 'Sharma', 19, 'M', 2000), ('Raj', 'Kandukuri', 20, 'M', 7000),
('Ramya', 'Ramapriya', 25, 'F', 5000),('Mac', 'Mohan', 26, 'M', 2000)]
cursor.executemany(insert_stmt, data)
conn.commit()

#Retrieving specific records using the where clause
cursor.execute("SELECT * from EMPLOYEE WHERE AGE <30")
print(cursor.fetchall())

#Closing the connection
conn.close()

[('Krishna', 'Sharma', 19, 'M', 2000.0), ('Raj', 'Kandukuri', 20, 'M', 7000.0), ('Ramya', 'Ramapriya', 25, 'F', 5000.0), ('Mac', 'Mohan', 26, 'M', 2000.0)]


Updating schemas using Python
(We update the ages of all the females by adding one year to their existing ages.)


In [3]:
import mysql.connector

conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Preparing the query to update the records
sql = '''UPDATE EMPLOYEE SET AGE = AGE + 1 WHERE SEX = 'F' '''
try:
   # Execute the SQL command
   cursor.execute(sql)
   
   # Commit your changes in the database
   conn.commit()
except:
   # Rollback in case there is any error
   conn.rollback()
   
#Retrieving data
sql = '''SELECT * from EMPLOYEE'''

#Executing the query
cursor.execute(sql)

#Displaying the result
print(cursor.fetchall())

#Closing the connection
conn.close()

[('Krishna', 'Sharma', 19, 'M', 2000.0), ('Raj', 'Kandukuri', 20, 'M', 7000.0), ('Ramya', 'Ramapriya', 26, 'F', 5000.0), ('Mac', 'Mohan', 26, 'M', 2000.0)]


### Task  (c) Add integrity constraints (Declare 'NotNull', Primary keys, Foreign keys). It should show errors if an integrity constraint can not be added. 

Adding integrity constraints

##### Let us assign the Unique constraint labelled 'con_unique' to the last_name column

In [4]:
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Preparing the query to update the records
sql = '''alter table employee add constraint con_unique Unique(Last_name) '''
try:
   # Execute the SQL command
   cursor.execute(sql)
   
   # Commit your changes in the database
   conn.commit()
except:
   # Rollback in case there is any error
   conn.rollback()
   
#Retrieving data
sql = '''desc employee'''

#Executing the query
cursor.execute(sql)

#Displaying the result
print(cursor.fetchall())

#Closing the connection
conn.close()

[('FIRST_NAME', b'char(20)', 'NO', '', None, ''), ('LAST_NAME', b'char(20)', 'YES', 'UNI', None, ''), ('AGE', b'int', 'YES', '', None, ''), ('SEX', b'char(1)', 'YES', '', None, ''), ('INCOME', b'float', 'YES', '', None, '')]


##### Since we have defined the Last_name column as unique, let us try inserting the same surname 'Singh' to another record to see if it gets inserted or not

In [23]:
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Preparing the query to update the records
sql = '''insert into employee values ('Archana' , 'Singh', 20, 'F', 6000, 6) '''

cursor.execute(sql)

myresult = cursor.fetchall()

for x in myresult:
  print(x)



#Closing the connection
conn.close()

IntegrityError: 1062 (23000): Duplicate entry 'Singh' for key 'employee.con_unique'

##### As can be seen from the 'IntegrityError' above : 1062 (23000): Duplicate entry 'Singh' for key 'employee.con_unique', 
The record with the same Last name 'Singh' could not be inserted owing to the integrity constraint of 'Unique' implemented in the previous python command.

In [24]:
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Preparing the query to update the records
sql = '''alter table employee add constraint Not null(income) '''
try:
   # Execute the SQL command
   cursor.execute(sql)
   
   # Commit your changes in the database
   conn.commit()
except:
   # Rollback in case there is any error
   conn.rollback()
   
#Retrieving data
sql = '''desc employee'''

#Executing the query
cursor.execute(sql)

#Displaying the result
print(cursor.fetchall())

#Closing the connection
conn.close()

[('FIRST_NAME', b'char(20)', 'NO', '', None, ''), ('LAST_NAME', b'char(20)', 'YES', 'UNI', None, ''), ('AGE', b'int', 'YES', '', None, ''), ('SEX', b'char(1)', 'YES', '', None, ''), ('INCOME', b'float', 'YES', '', None, ''), ('Emp_ID', b'int', 'YES', '', None, '')]


##### Now to assign foreign key and test referential integrity, we create another table Contact as shown below


In [5]:
#establishing the connection
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)

#Creating a cursor object using the cursor() method
cursor = conn.cursor()


#Creating table as per requirement
sql ='''CREATE TABLE CONTACT(
   Emp_ID INT NOT NULL,
   EMAIL CHAR(20) NOT NULL,
   PHONE LONG,
   CITY CHAR(20)
)'''
cursor.execute(sql)

#Closing the connection
#conn.close()

We also insert some records into the contact table for performing operations

In [ ]:

   #establishing the connection
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

# Preparing SQL query to INSERT a record into the database.
sql = """INSERT INTO CONTACT (Emp_ID, EMAIL, CITY) VALUES 
   (101, 'Krishna@mymail.com', 'Hyderabad'), 
   (102, 'Raja@mymail.com', 'Vishakhapatnam'), 
   (103, 'Krishna@mymail.com', 'Pune'), 
   (104, 'Raja@mymail.com', 'Mumbai')"""
   try:
   # Executing the SQL command
   cursor.execute(sql)

   # Commit your changes in the database
   conn.commit()

except:
   # Rolling back in case of error
   conn.rollback()


# Closing the connection
conn.close()


##### We now add the primary key constraint to the employee table

In [5]:
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Preparing the query to update the records
sql = '''alter table employee add constraint primary key(emp_ID) '''
try:
   # Execute the SQL command
   cursor.execute(sql)
   
   # Commit your changes in the database
   conn.commit()
except:
   # Rollback in case there is any error
   conn.rollback()
   
#Retrieving data
sql = '''desc employee'''

#Executing the query
cursor.execute(sql)

#Displaying the result
print(cursor.fetchall())

#Closing the connection
conn.close()

[('FIRST_NAME', b'char(20)', 'NO', '', None, ''), ('LAST_NAME', b'char(20)', 'YES', 'UNI', None, ''), ('AGE', b'int', 'YES', '', None, ''), ('SEX', b'char(1)', 'YES', '', None, ''), ('INCOME', b'float', 'YES', '', None, ''), ('Emp_ID', b'int', 'NO', 'PRI', None, '')]


In [6]:
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Preparing the query to update the records
sql = '''insert into employee values ('Kavya' , 'Deshmukh', 20, 'F', 7000, null) '''

cursor.execute(sql)

myresult = cursor.fetchall()

for x in myresult:
  print(x)



#Closing the connection
conn.close()

IntegrityError: 1048 (23000): Column 'Emp_ID' cannot be null

##### As can be seen, since the emp_Id column is a primary key, it does not accept a null value.
##### We now move on to enforce referential integrity. As can be seen from the schemas of the two tables, employee and contact, Emp_ID is the common column . Let us make the Emp_ID key of the employee table as the foreign key for our contact table. The necessary condition for this is that the emp_ID column of the employee table should be already assigned as the Primary key which we have done above

In [2]:
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Preparing the query to update the records
sql = '''alter table contact add Foreign key (emp_ID) references employee (emp_ID)'''
try:
   # Execute the SQL command
   cursor.execute(sql)
   
   # Commit your changes in the database
   conn.commit()
except:
   # Rollback in case there is any error
   conn.rollback()
   
#Retrieving data
sql = '''desc contact'''

#Executing the query
cursor.execute(sql)

#Displaying the result
print(cursor.fetchall())

#Closing the connection
conn.close()

[('Emp_ID', b'int', 'NO', '', None, ''), ('EMAIL', b'char(20)', 'NO', '', None, ''), ('PHONE', b'mediumtext', 'YES', '', None, ''), ('CITY', b'char(20)', 'YES', '', None, '')]


In [3]:
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Preparing the query to update the records
sql = '''insert into employee values ('Abhi' , 'Tiwari', 23, 'M', 4000, 6) '''

cursor.execute(sql)

myresult = cursor.fetchall()

for x in myresult:
  print(x)



#Closing the connection
conn.close()

In [5]:
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Preparing the query to update the records
sql = '''insert into contact values (null, 'abhi@gmail.com', 97602794902, 'Chennai') '''

cursor.execute(sql)

myresult = cursor.fetchall()

for x in myresult:
  print(x)



#Closing the connection
conn.close()

IntegrityError: 1048 (23000): Column 'Emp_ID' cannot be null

##### As can be seen from the above 'IntegrityError' message, we cannot have the foreign key column to be empty or null.

### Task  (d) Add/remove users and their privileges. It should show errors in case a privilege can not be given or removed. 

We will first create a new user

In [6]:
cnx = mysql.connector.connect(user='Rita', password='ainucurie@22',
                              host='127.0.0.1',
                              database='DSE310Rita')
cnx.close()

ProgrammingError: 1045 (28000): Access denied for user 'Rita'@'localhost' (using password: YES)

As can be seen, we had an access error, hence the below cell is the correct way to create a new user

In [8]:
#establishing the connection
conn = mysql.connector.connect(user='root', password='ainucurie@22', host='127.0.0.1')

#Creating a cursor object using the cursor() method
cursor = conn.cursor()

cursor.execute("CREATE USER 'Rita'@'localhost' IDENTIFIED BY 'rchaos@22';")


conn.close()

Now we attempt to create a new table named 'Student' in the same database with the new user 'Rita'

In [9]:
cnx = mysql.connector.connect(user='Rita', password='rchaos@22',
                              host='127.0.0.1',
                              database='DSE310Rita')

cursor = cnx.cursor()



#Creating table as per requirement
sql ='''CREATE TABLE student(
   FIRST_NAME CHAR(20) NOT NULL,
   LAST_NAME CHAR(20),
   AGE INT,
   SEX CHAR(1),
   MARKS FLOAT
)'''
cursor.execute(sql)
                              
cnx.close()

ProgrammingError: 1045 (28000): Access denied for user 'Rita'@'localhost' (using password: YES)

##### As can be seen we get the error : "ProgrammingError: 1045 (28000): Access denied for user 'Rita'@'localhost' (using password: YES)" since the newly created user does not have the necessary privileges to create schemas yet.

##### So now we grant all the privileges to the new user 'Rita' using the following commands given in the cell below 

In [35]:

#restablishing the connection
conn = mysql.connector.connect(
   user='root', password='ainucurie@22', host='127.0.0.1', database='DSE310Rita'
)
#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Retrieving single row
#sql = ''' GRANT CREATE, ALTER , DROP, INSERT, UPDATE, DELETE, SELECT, REFERENCES, RELOAD on DSE310Rita TO 'Rita'@'localhost' identified by 'rchaos@22' '''

#Executing the query
#cursor.execute(sql)
#cursor.execute("GRANT CREATE, ALTER, DROP, INSERT, UPDATE, DELETE, SELECT, REFERENCES, RELOAD on DSE310Rita.* TO 'Rita'@'localhost' ;")
cursor.execute("GRANT ALL PRIVILEGES ON DSE310Rita. * TO 'Rita'@'localhost';")


#Closing the connection
conn.close()

##### All the privileges have been granted to 'Rita' by the root host, so we can try to create schemas and alter them

In [37]:
cnx = mysql.connector.connect(user='Rita', password='rchaos@22',
                              host='127.0.0.1',
                              database='DSE310Rita')

cursor = cnx.cursor()



#Creating table as per requirement
sql ='''CREATE TABLE student(
   FIRST_NAME CHAR(20) NOT NULL,
   LAST_NAME CHAR(20),
   AGE INT,
   SEX CHAR(1),
   MARKS FLOAT
)'''
cursor.execute(sql)
                              
cnx.close()

In [39]:
cnx = mysql.connector.connect(user='Rita', password='rchaos@22',
                              host='127.0.0.1',
                              database='DSE310Rita')

cursor = cnx.cursor()

cursor.execute("desc student;")
myresult = cursor.fetchall()

for x in myresult:
  print(x)
                              
cnx.close()

('FIRST_NAME', b'char(20)', 'NO', '', None, '')
('LAST_NAME', b'char(20)', 'YES', '', None, '')
('AGE', b'int', 'YES', '', None, '')
('SEX', b'char(1)', 'YES', '', None, '')
('MARKS', b'float', 'YES', '', None, '')


In [42]:
cnx = mysql.connector.connect(user='Rita', password='rchaos@22',
                              host='127.0.0.1',
                              database='DSE310Rita')

cursor = cnx.cursor()

cursor.execute("insert into student values ('Riya', 'Gupta' , 20, 'F', 98);")
myresult = cursor.fetchall()

for x in myresult:
  print(x)
                              
cnx.close()

In [43]:
cnx = mysql.connector.connect(user='Rita', password='rchaos@22',
                              host='127.0.0.1',
                              database='DSE310Rita')

cursor = cnx.cursor()

cursor.execute("insert into student values ('Kalyanshi', 'sharma' , 20, 'F', 93);")
cursor.execute("insert into student values ('Saksham', 'Jain' , 21, 'M', 96);")
myresult = cursor.fetchall()

for x in myresult:
  print(x)
                              
cnx.close()

##### As can be seen from the above cell executions, the new user 'Rita' has created a new table , has also inserted values into the table

##### Let us now create another user, 'User2'  with lesser privileges and see how we can remove them.

In [44]:
#establishing the connection
conn = mysql.connector.connect(user='root', password='ainucurie@22', host='127.0.0.1')

#Creating a cursor object using the cursor() method
cursor = conn.cursor()

cursor.execute("CREATE USER 'user2'@'localhost' IDENTIFIED BY 'new@user';")


conn.close()

In [49]:
#establishing the connection
conn = mysql.connector.connect(user='root', password='ainucurie@22', host='127.0.0.1')

#Creating a cursor object using the cursor() method
cursor = conn.cursor()

cursor.execute("GRANT CREATE, ALTER, DROP, INSERT on DSE310Rita.*  TO 'user2'@'host' ;")


conn.close()

##### So we have successfully created another 'user2' with Create, alter, drop , and insert privileges.

In [50]:
cnx = mysql.connector.connect(user='user2', password='new@user',
                              host='127.0.0.1',
                              database='DSE310Rita')

cursor = cnx.cursor()

cursor.execute("update student set Marks = 93 where first_name = 'Riya';")

myresult = cursor.fetchall()

for x in myresult:
  print(x)
                              
cnx.close()

ProgrammingError: 1044 (42000): Access denied for user 'user2'@'localhost' to database 'dse310rita'

##### As can be seen above, since the 'User2' hasn't been given update privilege, he/she is unable to update any existing table in the database.

In [57]:
cnx = mysql.connector.connect(user='root', password='ainucurie@22',
                              host='127.0.0.1',
                              database='DSE310Rita')

cursor = cnx.cursor()

cursor.execute("REVOKE INSERT ON *.* FROM 'user2'@'localhost'; ")

myresult = cursor.fetchall()

for x in myresult:
  print(x)
                              
cnx.close()

##### Hence, the host or the root user has revoked the 'Insert' privilege that user2 had earlier. Let's test if the command has worked succesfully or not

In [58]:
cnx = mysql.connector.connect(user='user2', password='new@user',
                              host='127.0.0.1',
                              database='DSE310Rita')

cursor = cnx.cursor()

cursor.execute("insert into student values ('Kavya', 'Shah' , 19, 'F', 96);")

myresult = cursor.fetchall()

for x in myresult:
  print(x)
                              
cnx.close()

ProgrammingError: 1044 (42000): Access denied for user 'user2'@'localhost' to database 'dse310rita'

##### As can be seen from the error message above : "ProgrammingError: 1044 (42000): Access denied for user 'user2'@'localhost' to database 'dse310rita'"  , the access to user2 for inserting has been denied because we have droppped or revoked the insert privilege from that particular user.

Code submitted by Rita Abani 19244
Thanks and Regards
